assets/web/js/web.js

```diff
@@ -432,4 +330,287 @@ const $document = $(document),
+
+        $button.data('quantity', quantity)
+        $costField.text(`$${quantity * subPrice}`)
+    }
+
+    $('.selectable').click(function() {
+        selectSelectableElement(this)
+    })
+
+    $('.expansion-arrow').click(function() {
+        toggleExpandingElementForArrow(this)
+    })
+
+    $pricingOptionsSelectables.click(function() {
+        updateQuantityButtonCosts()
+    })
+
+    $searchUserForm.submit(function(event) {
+        event.preventDefault()
+
+        const username = $searchUserUsernameInput.val().trim()
+        $searchUserUsernameInput.val(username)
+        if (username === '') {
+            return
+        }
+
+        $.ajax({
+            url: '/api/info/giftcheck',
+            data: {s: username},
+            type: 'GET',
+            success: function(data) {
+                if (data['valid'] && data['cangift']) {
+                    setSearchUserSuccess('This user can accept gift subs!')
+                } else if (!data['valid'] && !data['cangift']) {
+                    setSearchUserError('This user doesn\'t exist.')
+                } else if (data['valid'] && !data['cangift']) {
+                    setSearchUserError('This user can\'t accept gift subs.')
+                } else if (!data['valid'] && data['cangift']) {
+                    setSearchUserError('YEE NEVA EVA LOSE.')
+                }
+            },
+            error: function(xhr) {
+                if (xhr.status === 403) {
+                    // This is required because of a server-side check that
+                    // prevents the user from gifting a sub to themselves.
+                    setSearchUserError('Sorry, you have to be logged in to search for users.')
+                } else {
+                    setSearchUserError('Something went wrong. Please try again later.')
+                }
+            }
+        })
+    })
+
+    $searchUserUsernameInput.keydown(function() {
+        clearSearchUserMessage()
+    })
+
+    $searchUserConfirmButton.click(function() {
+        confirmGiftee($searchUserUsernameInput.val())
+    })
+
+    $myselfSelectable.click(function() {
+        clearContinueFormErrorMessage()
+    })
+
+    $directGiftExpansionArrow.click(function() {
+        selectSelectableElement($directGiftSelectable)
+    })
+
+    $directGiftSelectable.click(function() {
+        toggleExpandingElementForArrow($directGiftExpansionArrow)
+    })
+
+    $massGiftSelectable.click(function() {
+        toggleExpandingElementForArrow($massGiftExpansionArrow)
+    })
+
+    $massGiftExpansionArrow.click(function() {
+        selectSelectableElement($massGiftSelectable)
+    })
+
+    $quantityInput.on('keyup change', function() {
+        // No negatives allowed.
+        if (event.type === 'keyup' && event.which === 189) {
+            return false
+        }
+
+        let quantity = parseInt($quantityInput.val())
+        if (isNaN(quantity)) {
+            return
+        }
+
+        if (quantity < 1) {

```

Comment:
this check for amount is fine here, but it also needs to be done server side. Someone could submit a value greater than 100 through a curl command, and the validation should exist on the server as well to constrain it.

Reply:
We actually do server-side validation for quantity limit and other conditions in [`SubscriptionController->validateSubscriptionParameters()`](https://github.com/destinygg/website/blob/66db093df2863efeeb01be5806cb3da454743e91/lib/Destiny/Controllers/SubscriptionController.php#L445-L477), which is used for [`/subscription/confirm`](https://github.com/destinygg/website/blob/66db093df2863efeeb01be5806cb3da454743e91/lib/Destiny/Controllers/SubscriptionController.php#L182) and [`/subscription/create`](https://github.com/destinygg/website/blob/66db093df2863efeeb01be5806cb3da454743e91/lib/Destiny/Controllers/SubscriptionController.php#L232).